File: 36/double_base_palindromes.py
#!/usr/local/bin/python3

from functools import reduce

def binary(n):
	return int(bin(n).lstrip('-0b'))

def is_palindrome(n):
	num_str = str(n)
	num_len = len(num_str)
	for ix,ch in enumerate(num_str):
		if ix >= num_len / 2:
			break
		if num_str[num_len - ix - 1] != ch:
			return False
	return True

double_palindromes = []

for i in range(1, 1000000):
	if is_palindrome(i) and is_palindrome(binary(i)):
		double_palindromes.append(i)

print("There are {} double palindromes with a total sum of {}.".format(
	len(double_palindromes),
	reduce(lambda x,y: x+y, double_palindromes)
))

File: 36/double_base_palindromes.rb
#!/usr/bin/ruby

def is_palindrome(n)
	num_str = n.to_s
	length = num_str.length
	num_str.split('').each_with_index do |ch,ix|
		return true if ix > length
		return false if num_str[length - 1 - ix] != ch
	end
	true
end

def binary(n)
	n.to_s(2).to_i
end

double_palindromes = []

(1..1000000).each do |i|
	double_palindromes.push(i) if is_palindrome(i) && is_palindrome(binary(i))
end

puts "There are #{double_palindromes.length} double palindromes with a total sum of #{double_palindromes.reduce(:+)}."
